### PR TITLE
CFD-372 Group approval workflow: Extend groups with extra field "status"

### PR DESCRIPTION
**Pull request by HelenaEksler**
#536

### CODE REVIEW

**Comment by HelenaEksler** [comment]
Update:
- Created field c4m_og_status (list)
- Handled field access through hook_field_access based on user role and og_role
- Handled field allowed values through hook_form_alter based on user role and og_role

File: capacity4more/modules/c4m/field/c4m_field_og/c4m_field_og.module [hunk 1]
```diff
@@ -28,3 +28,118 @@ function c4m_field_og_ds_fields_info($entity_type) {
...
+
+  $account = user_load($user->uid);
+  $entity = $form_state['node'];
+  $wrapper = entity_metadata_wrapper('node', $entity);
+  $value = $wrapper->c4m_og_status->value();
+
+  if ($account->uid == $entity->uid) {
+    // Group owner.
+
+    $allowed_values = array(
```
Comment (amitaibu) [inline capacity4more/modules/c4m/field/c4m_field_og/c4m_field_og.module]: $allowed values should just be the keys, and remove what's not needed.

File: capacity4more/modules/c4m/field/c4m_field_og/c4m_field_og.module [hunk 2]
```diff
@@ -28,3 +28,118 @@ function c4m_field_og_ds_fields_info($entity_type) {
...
+        'deleted' => 'Deleted',
+        'published' => 'Published',
+      ),
+      'published' => array(
+        'published' => 'Published',
+        'archived' => 'Archived',
+        'deleted' => 'Deleted',
+      ),
+    );
+    if (array_key_exists($value, $allowed_values)) {
```
Comment (amitaibu) [inline capacity4more/modules/c4m/field/c4m_field_og/c4m_field_og.module]: ``!empty()``

File: capacity4more/modules/c4m/field/c4m_field_og/c4m_field_og.module [hunk 3]
```diff
@@ -28,3 +28,118 @@ function c4m_field_og_ds_fields_info($entity_type) {
...
+      $form['c4m_og_status']['und']['#options'] = $allowed_values[$value];
```
Comment (amitaibu) [inline capacity4more/modules/c4m/field/c4m_field_og/c4m_field_og.module]: 'und' => LANGUAGE_NONE

File: capacity4more/modules/c4m/field/c4m_field_og/c4m_field_og.module [hunk 4]
```diff
@@ -28,3 +28,118 @@ function c4m_field_og_ds_fields_info($entity_type) {
...
+}
+
+/**
+ * Implements hook_form_alter().
+ */
+function c4m_field_og_form_alter(&$form, &$form_state, $form_id) {
+  if (!isset($form_state['field']['c4m_og_status'])) {
+    return;
+  }
+  global $user;
```
Comment (amitaibu) [inline capacity4more/modules/c4m/field/c4m_field_og/c4m_field_og.module]: first line

File: capacity4more/modules/c4m/field/c4m_field_og/c4m_field_og.module [hunk 5]
```diff
@@ -28,3 +28,118 @@ function c4m_field_og_ds_fields_info($entity_type) {
...
+    }
+  }
+
+  return FALSE;
+}
+
+/**
+ * Implements hook_form_alter().
+ */
+function c4m_field_og_form_alter(&$form, &$form_state, $form_id) {
```
Comment (amitaibu) [inline capacity4more/modules/c4m/field/c4m_field_og/c4m_field_og.module]: move to the specific ``hook_form_FORM_ID_alter()``

File: capacity4more/modules/c4m/field/c4m_field_og/c4m_field_og.module (hunk 1 above)
Comment (HelenaEksler) [inline capacity4more/modules/c4m/field/c4m_field_og/c4m_field_og.module]: Fixed in proper place.

**Comment by nader77** [comment]
@HelenaEksler Do you think we need to add this field to the demo migration to have a variety of statuses  ?

**Comment by HelenaEksler** [comment]
@zero2one ^^

**Comment by zero2one** [comment]
Yes add some different test statuses.

You could use them in the Behat tests to validate who can change what status.

**Comment by HelenaEksler** [comment]
selection_298

selection_297

**Comment by HelenaEksler** [comment]
@nader77 Could you please review it?

**Comment by nader77** [comment]
@HelenaEksler Looks good, Assign to Peter when test are green.

File: capacity4more/behat/features/group_status.feature [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+
+  @api
+  Scenario: Check requested group as admin
+    Given I am logged in as user "mariecurie"
+    When  I start editing group "Requested group"
+    Then  the "#edit-c4m-og-status-und" element should not contain "archived"
+    And   the "#edit-c4m-og-status-und" element should not contain "published"
+
+  @api
+  Scenario: Check requested group as group owner
```
Comment (zero2one) [inline capacity4more/behat/features/group_status.feature]: @HelenaEksler A group owner has no access to its group until it is reviewed and approved.

Comment (HelenaEksler) [inline capacity4more/behat/features/group_status.feature]: @zero2one But he can change status for requested to "deleted" due to schema. And due to the Group Rights table group owner can view group and change group properties (incl. group status -> to deleted as I understood).

File: capacity4more/behat/features/group_status.feature [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+    Given I am logged in as user "alfrednobel"
+    When  I start editing group "Requested group"
+    Then  the "#edit-c4m-og-status-und" element should not contain "rejected"
+    And   the "#edit-c4m-og-status-und" element should not contain "draft"
+    And   the "#edit-c4m-og-status-und" element should not contain "published"
+    And   the "#edit-c4m-og-status-und" element should not contain "archived"
+
+  @api
+  Scenario: Check rejected group group as admin
+    Given I am logged in as user "mariecurie"
```
Comment (zero2one) [inline capacity4more/behat/features/group_status.feature]: @HelenaEksler A group owner should get access denied when a group is requested, rejected or deleted

File: capacity4more/behat/features/group_status.feature [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+    When  I start editing group "Rejected group"
+    Then  the "#edit-c4m-og-status-und" element should not contain "requested"
+    And   the "#edit-c4m-og-status-und" element should not contain "draft"
+    And   the "#edit-c4m-og-status-und" element should not contain "published"
+    And   the "#edit-c4m-og-status-und" element should not contain "archived"
+
+  @api
+  Scenario: Check rejected group group as group owner
+    Given I am logged in as user "alfrednobel"
```
Comment (zero2one) [inline capacity4more/behat/features/group_status.feature]: @HelenaEksler A group owner should get access denied when a group is requested, rejected or deleted

File: capacity4more/behat/features/group_status.feature [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+  Scenario: Check archived group as admin
+    Given I am logged in as user "mariecurie"
+    When  I start editing group "Archived group"
+    Then  the "#edit-c4m-og-status-und" element should not contain "rejected"
+    And   the "#edit-c4m-og-status-und" element should not contain "requested"
+    And   the "#edit-c4m-og-status-und" element should not contain "draft"
+
+  @api
+  Scenario: Check archived group as group owner
+    Given I am logged in as user "alfrednobel"
```
Comment (zero2one) [inline capacity4more/behat/features/group_status.feature]: An archived group can no longer be edited by any group member (including owner & group admins).

File: capacity4more/behat/features/group_status.feature (hunk 8 above)
Comment (HelenaEksler) [inline capacity4more/behat/features/group_status.feature]: @zero2one mariecure is not a group owner - it's site admin. Group owner is alfrednobel and he doesn't see the field so he can't change it https://github.com/capacity4dev/capacity4more/pull/537/files#diff-d14bf0915f4fa490a2acd63d3184bbf2R31

File: capacity4more/behat/features/group_status.feature [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+  @api
+  Scenario: Check deleted group as admin
+    Given I am logged in as user "mariecurie"
+    When  I start editing group "Deleted group"
+    Then  the "#edit-c4m-og-status-und" element should not contain "rejected"
+    And   the "#edit-c4m-og-status-und" element should not contain "requested"
+
+  @api
+  Scenario: Check deleted group as group owner
+    Given I am logged in as user "alfrednobel"
```
Comment (zero2one) [inline capacity4more/behat/features/group_status.feature]: @HelenaEksler A group owner should get access denied when a group is requested, rejected or deleted.

File: capacity4more/modules/c4m/field/c4m_field_og/c4m_field_og.module [hunk 11]
```diff
@@ -28,3 +28,172 @@ function c4m_field_og_ds_fields_info($entity_type) {
 
   return $fields;
 }
+
+/**
+ * Implements hook_field_access().
+ */
+function c4m_field_og_field_access($op, $field, $entity_type, $entity, $account) {
```
Comment (zero2one) [inline capacity4more/modules/c4m/field/c4m_field_og/c4m_field_og.module]: @HelenaEksler Limit the items that are available depending on the current status and the possible statuses from there.

Not every user role can select any status option.

See https://wiki.amplexor.com/confluence/display/C4DNG/Group+approval+workflow

**Comment by HelenaEksler** [comment]
@zero2one All node access stuff is going to be done in the following PR of the issue CFD-373.

Is done here:
- Created field c4m_og_status
- Handled who can change the field and to which states. Due to the Status transition diagram [here](https://wiki.amplexor.com/confluence/display/C4DNG/Group+approval+workflow) 

 So this PR can be merged.

**Pull request merged**